File: crawl/bot.py
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin, urlparse
from collections import deque
import time
import mimetypes
import os
from pymongo import MongoClient

class SimpleWebCrawler:
    def __init__(self, base_url):
        self.base_url = base_url
        self.visited_urls = set()
        self.url_queue = deque([base_url])
        self.data = {}  # Dictionnaire pour stocker les URLs, leurs liens et contenus

        # Configurer la connexion à MongoDB
        self.client = MongoClient("mongodb+srv://kalisearch:12RJKw75ElO8dTUd@cluster0.z8zdf0m.mongodb.net/")
        self.db = self.client['kali_search']  # Nom de la base de données
        self.collection = self.db['web_data']  # Nom de la collection

        # Test de connexion MongoDB
        self.collection.insert_one({'test': 'Connexion établie'})

    def fetch_page(self, url):
        try:
            response = requests.get(url)
            response.raise_for_status()
            return response.text, response.headers.get('Content-Type')
        except requests.RequestException as e:
            print(f"Échec de la récupération de {url}: {e}")
            return None, None

    def parse_links(self, html, base_url):
        soup = BeautifulSoup(html, 'html.parser')
        links = set()
        for a_tag in soup.find_all('a', href=True):
            full_url = urljoin(base_url, a_tag['href'])
            if self.base_url in full_url:  # S'assurer de rester dans le domaine de base
                links.add(full_url)
        return links

    def extract_text(self, html):
        soup = BeautifulSoup(html, 'html.parser')
        # Extraire tout le texte visible et le retourner sous forme de lignes
        texts = soup.stripped_strings
        return "\n".join(texts)

    def crawl(self):
        while self.url_queue:
            current_url = self.url_queue.popleft()
            if current_url not in self.visited_urls:
                print(f"Visite de : {current_url}")
                html, content_type = self.fetch_page(current_url)
                if html is not None:
                    self.visited_urls.add(current_url)
                    links = self.parse_links(html, current_url)
                    # Vérifier si l'URL pointe vers un fichier
                    if content_type:
                        mime_type, _ = mimetypes.guess_type(current_url)
                        if mime_type and (mime_type.startswith('image') or mime_type.startswith('text')):
                            file_name = os.path.basename(urlparse(current_url).path)
                            content = file_name
                        else:
                            content = self.extract_text(html)
                    else:
                        content = self.extract_text(html)

                    # Stocker les liens et le contenu de la page dans MongoDB
                    try:
                        self.collection.insert_one({
                            'url': current_url,
                            'links': list(links),
                            'content': content
                        })
                        print(f"Document inséré pour {current_url}")
                    except Exception as e:
                        print(f"Échec de l'insertion dans MongoDB pour {current_url}: {e}")
                    
                    for link in links:
                        if link not in self.visited_urls:
                            self.url_queue.append(link)
                time.sleep(1)  # Être poli et éviter de surcharger le serveur

if __name__ == "__main__":
    base_url = "https://caritasdegoma.org/"
    crawler = SimpleWebCrawler(base_url)
    crawler.crawl()

File: crawl/bot.js
import axios from 'axios';
import cheerio from 'cheerio';
import { URL } from 'url';
import { MongoClient } from 'mongodb';
import mime from 'mime-types';
import path from 'path';

class SimpleWebCrawler {
    constructor(baseUrls) {
        this.baseUrls = baseUrls;
        this.visitedUrls = new Set();
        this.urlQueue = [...baseUrls];

        // Configurer la connexion à MongoDB
        const uri = "mongodb+srv://kalisearch:12RJKw75ElO8dTUd@cluster0.z8zdf0m.mongodb.net/";
        this.client = new MongoClient(uri, { useNewUrlParser: true, useUnifiedTopology: true });
        this.db = null;
        this.collection = null;
    }

    async init() {
        try {
            await this.client.connect();
            this.db = this.client.db('kali_search'); // Nom de la base de données
            this.collection = this.db.collection('web_data'); // Nom de la collection

            // Test de connexion MongoDB
            await this.collection.insertOne({ test: 'Connexion établie' });
            console.log("Connexion MongoDB établie.");
        } catch (error) {
            console.error("Erreur de connexion à MongoDB:", error);
        }
    }

    async fetchPage(url) {
        try {
            const response = await axios.get(url);
            return { data: response.data, contentType: response.headers['content-type'] };
        } catch (error) {
            console.error(`Échec de la récupération de ${url}:`, error);
            return { data: null, contentType: null };
        }
    }

    parseLinks(html, baseUrl) {
        const $ = cheerio.load(html);
        const links = new Set();
        $('a[href]').each((index, element) => {
            let href = $(element).attr('href');
            if (!href.includes('#')) {  // Filtrer les liens avec des fragments
                const fullUrl = new URL(href, baseUrl).href;
                if (this.baseUrls.some(base => fullUrl.startsWith(base))) {
                    links.add(fullUrl);
                }
            }
        });
        return links;
    }

    extractText(html) {
        const $ = cheerio.load(html);
        return $('body').text().trim().replace(/\s+/g, ' ');
    }

    async crawl() {
        while (this.urlQueue.length > 0) {
            const currentUrl = this.urlQueue.shift();
            if (!this.visitedUrls.has(currentUrl)) {
                console.log(`Visite de : ${currentUrl}`);
                const { data, contentType } = await this.fetchPage(currentUrl);
                if (data) {
                    this.visitedUrls.add(currentUrl);
                    const links = this.parseLinks(data, currentUrl);
                    let content;

                    if (contentType) {
                        const mimeType = mime.lookup(currentUrl);
                        if (mimeType && (mimeType.startsWith('image') || mimeType.startsWith('text'))) {
                            const fileName = path.basename(new URL(currentUrl).pathname);
                            content = fileName;
                        } else {
                            content = this.extractText(data);
                        }
                    } else {
                        content = this.extractText(data);
                    }

                    // Vérifier si l'URL est déjà indexée et mettre à jour ou insérer le document
                    try {
                        const existingDocument = await this.collection.findOne({ url: currentUrl });
                        if (existingDocument) {
                            await this.collection.updateOne(
                                { url: currentUrl },
                                { $set: { links: Array.from(links), content: content } }
                            );
                            console.log(`Document mis à jour pour ${currentUrl}`);
                        } else {
                            await this.collection.insertOne({
                                url: currentUrl,
                                links: Array.from(links),
                                content: content
                            });
                            console.log(`Document inséré pour ${currentUrl}`);
                        }
                    } catch (error) {
                        console.error(`Échec de l'insertion ou de la mise à jour dans MongoDB pour ${currentUrl}:`, error);
                    }

                    links.forEach(link => {
                        if (!this.visitedUrls.has(link)) {
                            this.urlQueue.push(link);
                        }
                    });
                }
                await new Promise(resolve => setTimeout(resolve, 1000)); // Être poli et éviter de surcharger le serveur
            }
        }
    }
}

(async () => {
    const baseUrls = [
        "https://elttec-africa.com/",
        "https://www.kadea.academy/",
        "https://goma-innovation.com/"
    ];
    const crawler = new SimpleWebCrawler(baseUrls);
    await crawler.init();
    await crawler.crawl();
    await crawler.client.close(); // Fermer la connexion MongoDB à la fin
})();
